pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -158,6 +158,45 @@ protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutab
     }
   }
 
+  /**
+   * <li> When the replacing segment and current segment are of {@link LLCSegmentName} then the PK should resolve to
+   * row in segment with higher sequence id.
+   * <li> When the replacing segment and current segment are of {@link UploadedRealtimeSegmentName} then the PK
+   * should resolve to row in segment with higher creation time
+   * <li> For other cases resolve based on creation time of segment. In case the creation time is same, give
+   * preference to an uploaded segment. A segment which is not LLCSegment can be assumed to be uploaded segment and
+   * is given preference.
+   *
+   * @param segmentName replacing segment name
+   * @param currentSegmentName current segment name having the record for the given primary key
+   * @param segmentCreationTimeMs replacing segment creation time
+   * @param currentSegmentCreationTimeMs current segment creation time
+   * @return true if the record in replacing segment should replace the record in current segment
+   */
+  protected boolean shouldReplaceOnComparisonTie(String segmentName, String currentSegmentName,
+      long segmentCreationTimeMs, long currentSegmentCreationTimeMs) {
+
+    LLCSegmentName llcSegmentName = LLCSegmentName.of(segmentName);
+    LLCSegmentName currentLLCSegmentName = LLCSegmentName.of(currentSegmentName);
+    if (llcSegmentName != null && currentLLCSegmentName != null) {
+      return llcSegmentName.getSequenceNumber() > currentLLCSegmentName.getSequenceNumber();
+    }
+
+    int creationTimeComparisonRes = Long.compare(segmentCreationTimeMs, currentSegmentCreationTimeMs);
```

Comment:
looks like we can do
```
...
if (llcSegmentName != null && currentLLCSegmentName != null) {
  return llcSegmentName.getSequenceNumber() > currentLLCSegmentName.getSequenceNumber();
}
// If any one is not LLC, then compare ctime
// likely ctime is not 0 so we save parsing cost of UploadedRealtimeSegmentName
if (creationTimeComparisonRes != 0) {
  return creationTimeComparisonRes > 0;
}
// their ctimes are same, then favor uploaded segment, no need to parse just yet
if (currentLLCSegmentName != null) {
  return  true; 
}

// both segments are uploaded, then prefer a well formatted name
// also favor the first writer in case both are well formatted
if (UploadedRealtimeSegmentName.of(currentSegmentName) != null) {
  return false;
}
if (UploadedRealtimeSegmentName.of(currentSegmentName) != null) {
  return true;
}
// favor the first writer
return false;
```
wdyt?

Reply:
> // favor the first writer
> return false;

This may be an inconsistent behaviour(there can be two writers and may not always sync and write) but I don't have a strong inclination. It will be good to keep it as you suggested as it does not modify the existing behaviour 

For later, we can always depend on a critera based on segment metadata/name to be consistent. e.g. we can check based on lexical order of segment name which can enforce to use right naming conventions.